data/src/test/java/org/oppia/android/data/backends/gae/NetworkLoggingInterceptorTest.kt

```diff
@@ -0,0 +1,240 @@
+    runBlockingTest {
+      val pageNotFound = HttpURLConnection.HTTP_NOT_FOUND
+
+      val mockResponse = MockResponse()
+        .setResponseCode(pageNotFound)
+        .setBody(testResponseBody)
+
+      mockWebServer.enqueue(mockResponse)
+
+      val networkJob = launch {
+        networkLoggingInterceptor.logNetworkCallFlow.collect {
+          assertThat(it.requestUrl).isEqualTo(mockWebServerUrl.toString())
+          assertThat(it.responseStatusCode).isEqualTo(pageNotFound)
+          assertThat(it.headers).contains(headerString)
+          assertThat(it.body).isEqualTo(testResponseBody)
+        }
+      }
+
+      val failedNetworkJob = launch {
+        networkLoggingInterceptor.logFailedNetworkCallFlow.collect {
+          assertThat(it.requestUrl).isEqualTo(mockWebServerUrl.toString())
+          assertThat(it.responseStatusCode).isEqualTo(pageNotFound)
+          assertThat(it.headers).contains(headerString)
+          assertThat(it.body).isEmpty()
+          assertThat(it.errorMessage).isNotEmpty()
+          assertThat(it.errorMessage).isEqualTo(testResponseBody)
+        }
+      }
+
+      client.newCall(request).execute()
+      testCoroutineDispatchers.advanceUntilIdle()
+      networkJob.cancel()
+      failedNetworkJob.cancel()
+    }
+
+  @Test
+  fun testLoggingInterceptor_makeFailingCallToTopicService_logsNetworkCallFailed_withException() =
+    runBlockingTest {
+      mockWebServer.shutdown()
+
+      val failedNetworkJob = launch {
+        networkLoggingInterceptor.logFailedNetworkCallFlow.collect {
+          assertThat(it.requestUrl).isEqualTo(mockWebServerUrl.toString())
+          assertThat(it.responseStatusCode).isEqualTo(0)
+          assertThat(it.headers).contains(headerString)
+          assertThat(it.body).isEmpty()
+          assertThat(it.errorMessage).isNotEmpty()
+          assertThat(it.errorMessage).contains("Failed to connect to localhost")
+        }
+      }
+
+      try {
+        client.newCall(request).execute()
+      } catch (e: Exception) { }
+
+      testCoroutineDispatchers.advanceUntilIdle()
+      failedNetworkJob.cancel()
+    }
+
+  private fun setUpTestApplicationComponent() {
+    ApplicationProvider.getApplicationContext<TestApplication>().inject(this)
+  }
+
+  private fun setUpRetrofit() {
+    mockWebServer = MockWebServer()
+    client = OkHttpClient.Builder()
+      .addInterceptor(networkLoggingInterceptor)
+      .build()
+
+    // Use retrofit with the MockWebServer here instead of MockRetrofit so that we can verify that
+    // the full network request properly executes. MockRetrofit and MockWebServer perform the same
+    // request mocking in different ways and we want to verify the full request is executed here.
+    // See https://github.com/square/retrofit/issues/2340#issuecomment-302856504 for more context.
+    retrofit = Retrofit.Builder()
+      .baseUrl(mockWebServer.url(testUrl))
```

Comment:
Using the mockwebserver url allows it to intercept network calls and return the expected response.

Reply:
But the mock webserver url == test url.
mockWebserver.url("") just sets the value of the mock webserver URL,  AFAIK. So this isn't actually intercepting the network calls.

Reply:
When I use `mockWebserver.url()`, it returns a url similar to `http://localhost:54086/` that contains a link to the local mock webserver client (and on the appropriate port). If I used the url directly, the fake network call would fail.